### ISSUE TITLE
Interactive case splitting generates name clashing with out-of-scope constructor

### ISSUE THREAD

**Issue by nad** [agda/agda#2522]
Consider the following code:
```agda
open import Agda.Builtin.Size

record R (A : Size → Set) (i : Size) : Set where
  field
    force : (j : Size< i) → A j

data D (A : Size → Set) (i : Size) : Set where
  c : R A i → D A i

postulate
  P : (A : Size → Set) → D A ω → D A ω → Set

F : (Size → Set) → Set
F A = (x : A ω) (y : D A ω) →
      P _ (c (record { force = λ j → x })) y
```
Agda rejects this code with the following error message:
```
x != ω of type Size
when checking that the expression y has type D (λ _ → A ω) ω
```
Note that `x` is not a size. At first I thought that some internal invariant had been broken, but then I tried renaming `x` to something else, and got the following error message:
```
x != ω of type Size
when checking that the expression y has type D (λ _ → A ω) ω
```
It seems as if Agda itself invents the name `x` (I checked that `x` is not mentioned in `Agda.Builtin.Size`). I think this is rather confusing.

This is a regression. With Agda 2.5.1.1 I get the following, more reasonable error message:
```
x₁ != ω of type Size
when checking that the expression y has type D (λ _ → A ω) ω
```

**Comment by nad** [comment]
Bisection suggests that this regression was introduced by @UlfNorell in 67d50a9159147f0aa990aef4ae3ccbbfdd2b56df:
```
[ fixed #2018 ] don't rename variables when adding them to the context

... but make sure to pick fresh names when the system is choosing the name.
```

**Comment by nad** [comment]
See #3053 for another test case.

**Comment by andreasabel** [comment]
Here is an example involving case-splitting:
```agda
open import Agda.Builtin.Bool

test : (true : Bool) → Set
test = {!!}  -- C-c C-c

-- Generates
--    test true = ?
-- and error:
-- Incomplete pattern matching for test. Missing cases:
--   test false
-- when checking the definition of test
```

**Comment by nad** [comment]
Here is a related problem:
```agda
module _ where

module M where

  data D : Set where
    c : D

open M using (D)

f : (c : D) → D
f = {!!}
```
If I "split on the result", then I get the following code:
```agda
f c = ?
```
This code is rejected with the following error message:
```
The pattern variable c has the same name as the constructor D.c
when checking the clause f c = ?
```

**Comment by jespercockx** [comment]
#3446 fixes these problems except for the last one by @nad . My new mechanism tries to avoid picking names that are already in scope, but as `c` is not in scope it's not avoided.

**Comment by jespercockx** [comment]
I'm not going to fix the remaining issue before the release.

**Comment by nad** [comment]
The remaining problem exists also in Agda 2.5.1.1, so the "regression in 2.5.2" label does not apply to this problem.

**Comment by barrettj12** [comment]
I'm not sure if this is the same issue as @nad mentioned [here](https://github.com/agda/agda/issues/2522#issuecomment-389133965), but when I run the following:
```agda
data _×_ (A B : Set) : Set where
  [_,_] : (a : A) → (b : B) → A × B

variable
  a b : Set

f : a × b → Set
f k = ?
```
and try to case-split on `k`, it gives
```agda
f [ a , b ] = ?
```
despite the fact that `a` is already defined in the context.

Tested with Agda v2.6.2 on Windows 10.

**Comment by andreasabel** [comment]
@barrettj12: Yes, this fits here, only that in your case the case-splitter does not produce faulty code: `f [ a , b ] = ?` parses fine as pattern variables may shadow definitions and other local variables.  The product is faulty if one of the pattern variables shadow a _constructor_ that is in scope or could be in scope if qualified.  (The rationale for the latter is that one might accidentially turn a pattern variable into a constructor-match by changing an import statement.)

@jespercockx:  Maybe you can comment on what your design aims to achieve wrt. shadowing.

**Comment by barrettj12** [comment]
Right, yeah it didn't cause a compilation error, since it just renamed `a b : Set` to `a₁ b₁ : Set` in the local context. Still, I don't like Agda deciding to rename things that I've already given a name to - I think it would be better if the subscript 1's went on the new arguments, i.e. `f [ a₁ , b₁ ] = ?`

**Comment by jespercockx** [comment]
It is an intentional design choice to not consider names of generalizable variables when picking names for pattern variables. For other definitions Agda does pick a different name, e.g. if you turn `a` and `b` into postulates you get
```agda
f : a × b → Set
f [ a₁ , b₁ ] = {!!}
```
I actually think it makes sense in the case of `variables` that you'd want to shadow them in the pattern, since they can anyway not be used in the RHS (only in type signatures).

**Comment by andreasabel** [comment]
The case splitting example https://github.com/agda/agda/issues/2522#issuecomment-389133965 is still open.
